packages/sdk/android/src/main/java/com/primerioreactnative/components/manager/ach/PrimerRNHeadlessUniversalCheckoutStripeAchUserDetailsComponent.kt

```diff
@@ -0,0 +1,310 @@
+              JSONObject().apply {
+                putData(validationStatus.collectableData as AchUserDetailsCollectableData)
+              }
+          )
+        }
+        is PrimerValidationStatus.Invalid -> {
+          sendEvent(
+              PrimerHeadlessUniversalCheckoutComponentEvent.ON_IN_VALID.eventName,
+              JSONObject().apply {
+                putData(validationStatus.collectableData as AchUserDetailsCollectableData)
+                put(
+                    "errors",
+                    JSONArray(
+                        validationStatus.validationErrors.map {
+                          JSONObject(
+                              json.encodeToString(
+                                  PrimerValidationErrorRN(
+                                      it.errorId,
+                                      it.description,
+                                      it.diagnosticsId,
+                                  )
+                              )
+                          )
+                        }
+                    )
+                )
+              }
+          )
+        }
+        is PrimerValidationStatus.Valid -> {
+          sendEvent(
+              PrimerHeadlessUniversalCheckoutComponentEvent.ON_VALID.eventName,
+              JSONObject().apply {
+                putData(validationStatus.collectableData as AchUserDetailsCollectableData)
+              }
+          )
+        }
+        is PrimerValidationStatus.Error -> {
+          sendEvent(
+              PrimerHeadlessUniversalCheckoutComponentEvent.ON_VALIDATION_ERROR.eventName,
+              JSONObject().apply {
+                putData(validationStatus.collectableData as AchUserDetailsCollectableData)
+                put(
+                    "errors",
+                    JSONArray().apply {
+                      put(JSONObject(json.encodeToString(validationStatus.error.toPrimerErrorRN())))
+                    }
+                )
+              }
+          )
+        }
+      }
+    }
+  }
+
+  private fun JSONObject.putData(collectableData: AchUserDetailsCollectableData) {
+    put(
+        "data",
+        JSONObject(
+            json.encodeToString(
+                when (collectableData) {
+                  is AchUserDetailsCollectableData.FirstName ->
+                      collectableData.toFirstNameRN()
+                  is AchUserDetailsCollectableData.LastName -> collectableData.toLastNameRN()
+                  is AchUserDetailsCollectableData.EmailAddress ->
+                      collectableData.toEmailAddressRN()
+                }
+            )
+        )
+    )
+  }
+
+  @ReactMethod fun addListener(eventName: String?) = Unit
+
+  @ReactMethod fun removeListeners(count: Int?) = Unit
+
+  private fun JSONObject?.toWritableMap(): WritableMap =
+      this?.let { convertJsonToMap(this) } ?: Arguments.createMap()
+
+  private fun JSONArray?.toWritableArray(): WritableArray =
```

Comment:
[nit]: we seem to have this in every component, maybe worth centralising this and use in all components?

Reply:
Will do.